### PR TITLE
Improved CMake project

### PR DESCRIPTION
**Pull request by podsvirov**
Hi guys!
You are here Betta released! That's great, but where the binaries for C ++ for Windows?

I suggest a couple of commits to improve gusto project.

I tried not to change a lot, but some questions are very important:
- Improved parsing of the configuration script
- Search ZLIB the same for all platforms
- Using ZLIB target imported, if applicable

Hopefully this PR will be accepted quickly.

Betta these changes for Windows can be downloaded via the testing
installers of my project "DaD's House":

http://dad.podsvirov.pro

### CODE REVIEW

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -4,39 +4,58 @@ cmake_minimum_required(VERSION 2.8)
 # Project
 project(protobuf C CXX)
 
+# CMake policies
+cmake_policy(SET CMP0022 NEW)
+
 # Options
+option(protobuf_VERBOSE "Enable for verbose output" OFF)
```
Comment (xfxyjwf) [inline cmake/CMakeLists.txt]: How about just use "VERBOSE"? We should make these options named consistently.

File: cmake/CMakeLists.txt [hunk 2]
```diff
@@ -44,18 +63,17 @@ if (CMAKE_USE_PTHREADS_INIT)
...
-  if (ZLIB)
-    set(HAVE_ZLIB 1)
-    find_path(ZLIB_INCLUDE_DIRECTORIES zlib.h ${protobuf_SOURCE_DIR})
-    find_library(ZLIB_LIBRARIES zdll ${protobuf_SOURCE_DIR})
-  else (ZLIB)
-    set(HAVE_ZLIB 0)
-  endif (ZLIB)
-else (MSVC)
+if (ZLIB)
   find_package(ZLIB)
```
Comment (xfxyjwf) [inline cmake/CMakeLists.txt]: I don't know how this simplified version is going to work with MSVC build with ZLIB. Could you update the instructions in cmake/README.md about the zlib build part in this pull request?

File: cmake/CMakeLists.txt (hunk 1 above)
Comment (podsvirov) [inline cmake/CMakeLists.txt]: This is not a good practice to use short variable names for the project settings.
Usually VERBOSE option is only needed to debug a specific project.
If this project is included in the build process of a larger project, and it is quite possible with Git and CMake, then the settings of the current project will be coordinated with the policies of other project? ..
I would also like to rename the option ZLIB in protobuf_WITH_ZLIB. What do you think about it?

File: cmake/CMakeLists.txt (hunk 2 above)
Comment (podsvirov) [inline cmake/CMakeLists.txt]: I do not feel any difficulty on the Windows with MSVC.
You can easily find pre-builded and installed version of the ZLIB library with a standard CMake module FindZLIB.cmake.
To find the library, you can use standard hints:
cmake -D[CMAKE_PREFIX_PATH](http://www.cmake.org/cmake/help/v3.3/variable/CMAKE_PREFIX_PATH.html?highlight=cmake_prefix_path)=<path_to_zlib_install>
(cmake -D[CMAKE_INSTALL_PREFIX](http://www.cmake.org/cmake/help/v3.3/variable/CMAKE_INSTALL_PREFIX.html?highlight=cmake_install_prefix)=<path_to_zlib_install>)
Or more specific hint:
  cmake -DZLIB_ROOT=<path_to_zlib_install>
See more in module documentation [FindZLIB](http://www.cmake.org/cmake/help/v3.3/module/FindZLIB.html).

File: cmake/CMakeLists.txt [hunk 3]
```diff
@@ -4,39 +4,58 @@ cmake_minimum_required(VERSION 2.8)
...
 option(BUILD_TESTING "Build tests" ON)
 option(BUILD_SHARED_LIBS "Build Shared Libraries" OFF)
+option(protobuf_MSVC_STATIC_RUNTIME "Link static runtime libraries" ON)
 if (MSVC)
-  option(ZLIB "Build with zlib support" OFF)
+  set(ZLIB_DEFAULT OFF)
+else (MSVC)
+  set(ZLIB_DEFAULT ON)
 endif (MSVC)
+option(ZLIB "Build with zlib support" ${ZLIB_DEFAULT})
```
Comment (podsvirov) [inline cmake/CMakeLists.txt]: We can rename this option from <b>ZLIB</b> to <b>protobuf_WITH_ZLIB</b>?

File: cmake/CMakeLists.txt (hunk 1 above)
Comment (xfxyjwf) [inline cmake/CMakeLists.txt]: I would be fine if all the options are named consistently (i.e., all without prefix or all with prefix). If project_OPTION_NAME is a common convention in cmake, using it is fine to me. I know that gtest uses the simple form though. Could you show me some references to projects using the protobuf_VERBOSE style naming?

File: cmake/CMakeLists.txt (hunk 2 above)
Comment (xfxyjwf) [inline cmake/CMakeLists.txt]: Could you update the cmake/README.md file to include such instructions?

File: cmake/CMakeLists.txt (hunk 1 above)
Comment (podsvirov) [inline cmake/CMakeLists.txt]: You can see the [CMakeLists.txt](https://github.com/Kitware/CMake/blob/master/CMakeLists.txt) file of the [CMake](http://www.cmake.org) build tool. For example line [63](https://github.com/Kitware/CMake/blob/master/CMakeLists.txt#L63) option *CMake_INSTALL_COMPONENTS* of line [74](https://github.com/Kitware/CMake/blob/master/CMakeLists.txt#L74) option *CMake_INSTALL_DEPENDENCIES*. These options have been created on my initiative.
Then we will continue to use the *project_OPTION_NAME* variable naming.

File: cmake/CMakeLists.txt (hunk 2 above)
Comment (podsvirov) [inline cmake/CMakeLists.txt]: The instructions may vary depending on platform and compiler, as well as using only the console or IDE. On what should focus to begin with. I understand that now is only about peq building on Windows with MSVC?

File: cmake/CMakeLists.txt (hunk 1 above)
Comment (xfxyjwf) [inline cmake/CMakeLists.txt]: Thanks. Using project_OPTION_NAME for all options seems good to me.

File: cmake/CMakeLists.txt (hunk 2 above)
Comment (xfxyjwf) [inline cmake/CMakeLists.txt]: Yes, the cmake/README.md only talks about MSVC build and we only need instructions on MVSC build for the time being.

Comment (podsvirov) [inline cmake/CMakeLists.txt]: I also noticed that the file does not contain installation instructions, although it stated.

**Comment by podsvirov** [comment]
Well. After renaming CMake project options СI outraged .. But I can begin work on updating cmake/README.md file.

**Comment by podsvirov** [comment]
If these changes are adopted, then here #822 PR is for documentation updates.

**Comment by podsvirov** [comment]
If these changes are adopted, then here #822 PR is for documentation updates.

File: cmake/tests.cmake [hunk 4]
```diff
@@ -2,6 +2,9 @@ if (NOT EXISTS "${PROJECT_SOURCE_DIR}/../gmock/CMakeLists.txt")
   message(FATAL_ERROR "Cannot find gmock directory.")
 endif()
 
+option(protobuf_ABSOLUTE_TEST_PLUGIN_PATH
+  "Using absolute test_plugin path in tests" ON)
```
Comment (xfxyjwf) [inline cmake/tests.cmake]: Is this needed for "nmake check"?

Comment (podsvirov) [inline cmake/tests.cmake]: Yes.

**Comment by podsvirov** [comment]
Hello @xfxyjwf.
Any news on this PR?

**Comment by xfxyjwf** [comment]
appveyor tests are failing for this PR. Could you fix that?
https://ci.appveyor.com/project/protobuf/protobuf/build/1.0.585/job/lrtsvwkluv6owsot

**Comment by podsvirov** [comment]
Now AppVeyour Build success:
https://ci.appveyor.com/project/protobuf/protobuf/build/1.0.707/job/rqc200c75itkjr5n
@xfxyjwf more questions?

**Comment by podsvirov** [comment]
Excellent. I'm having some issues with zlib. I plan to do it soon.

**Comment by podsvirov** [comment]
Oops. Sorry. It seems that now everything is in order.

**Pull request merged**